watches.go

```diff
@@ -52,39 +67,72 @@ func (wm *watchManager) handleWatchRequest(w http.ResponseWriter, r *http.Reques
 		http.Error(w, fmt.Sprintf("invalid 'fromIndex': %v", err), http.StatusBadRequest)
 		return
 	}
-
-	go wm.manageWatch(conn, topic, genID, idx)
+	log.Printf("Connection accepted from %v", r.Host)
+	err = wm.manageWatch(w, topic, genID, idx)
+	if err != nil {
+		log.Printf("Error: handling messages %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 }
 
-func (wm *watchManager) manageWatch(conn *websocket.Conn, topic, genID string, idx uint64) {
-	log.Printf("Connection accepted from %v", conn.RemoteAddr())
-	defer closeConn(conn)
-	for {
-		msgsResponse, err := wm.store.get(topic, genID, idx)
-		if err != nil {
-			handleError(err, conn)
-			return
+func (wm *watchManager) manageWatch(w http.ResponseWriter, topic string, genID string, idx uint64) error {
+	var err error
+	var closed bool
+	cn, canNotifyClose := w.(http.CloseNotifier)
+	flusher, canFlush := w.(http.Flusher)
+	if canNotifyClose && canFlush {
+		for aw := newActiveWatch(wm, topic, genID, idx, httputil.NewChunkedWriter(w), flusher); ; {
+			select {
+			case <-cn.CloseNotify():
+				aw.close()
+			default:
+				err = aw.handleNewMessages()
+			}
+			if closed || err != nil {
+				break
+			}
+			time.Sleep(wm.pushInterval)
 		}
+	} else {
+		err = errors.New("Error: cannot stream")
+	}
+	return err
+}
+
+func (aw *activeWatch) handleNewMessages() error {
+	var err error
+	var msgsResponse *MessagesResponse
+	if msgsResponse, err = aw.newMessages(); err == nil {
 		if msgsLength := len(msgsResponse.Messages); msgsLength > 0 {
-			if err := conn.WriteJSON(msgsResponse); err != nil {
-				handleError(err, conn)
-				return
+			if err := aw.writeChunk(msgsResponse); err == nil {
+				aw.genID = msgsResponse.GenerationID
+				aw.idx = msgsResponse.Messages[msgsLength-1].Index + 1
 			}
-			genID = msgsResponse.GenerationID
-			idx = msgsResponse.Messages[msgsLength-1].Index + 1
 		}
-		time.Sleep(wm.pushInterval)
 	}
+	return err
+}
+
+func (aw *activeWatch) newMessages() (*MessagesResponse, error) {
+	return aw.wm.store.get(aw.topic, aw.genID, aw.idx)
 }
 
-func closeConn(conn *websocket.Conn) {
-	log.Printf("Terminating connection to %v", conn.RemoteAddr())
-	if err := conn.Close(); err != nil {
-		log.Printf("[WARNING] error closing connection: %v", err)
+func (aw *activeWatch) writeChunk(msgs *MessagesResponse) error {
+	var marshalled []byte
+	var err error
+	if marshalled, err = json.Marshal(msgs); err == nil {
+		if _, err = aw.cw.Write(marshalled); err == nil {
+			aw.flusher.Flush()
+		}
 	}
+	return err
```

Comment:
nit: the error checking here isn't idiomatic go. use `return` statements instead of nesting your error checks like so:
```go

func (aw *activeWatch) writeChunk(msgs *MessagesResponse) error {
	marshalled, err := json.Marshal(msgs)
	if err != nil {
		return err
	}
	if _, err = aw.cw.Write(marshalled); err != nil {
		return err
	}
	return aw.flusher.Flush()
}
```